clang/test/CodeGenSYCL/regcall-cc-test.cpp

```diff
@@ -0,0 +1,360 @@
+// CHECK-DAG: declare x86_regcallcc <8 x float> @_Z17__regcall3__SCALE4simdIfLi8EE(<8 x float>)
+
+SYCL_DEVICE simd<float, 8> __regcall foo(simd<float, 8> x) {
+  return SCALE(x);
+// CHECK-DAG: %{{[0-9a-zA-Z_.]+}} = call x86_regcallcc <8 x float> @_Z17__regcall3__SCALE4simdIfLi8EE(<8 x float> %{{[0-9a-zA-Z_.]+}})
+}
+
+// === TEST CASE: nested struct with different types of fields
+
+struct C {
+  float x, y;
+};
+// CHECK-DAG: %struct.C = type { float, float }
+
+struct PassAsByval {
+  C a;
+  int *b;
+  raw_vector<float, 3> c;
+};
+// CHECK-DAG: %struct.PassAsByval = type { %struct.C, i32 addrspace(4)*, <3 x float> }
+
+SYCL_DEVICE PassAsByval __regcall bar(PassAsByval x) {
+// CHECK-DAG: define dso_local x86_regcallcc %struct.PassAsByval @_Z15__regcall3__bar11PassAsByval(%struct.C %{{[0-9a-zA-Z_.]+}}, i32 addrspace(4)* %{{[0-9a-zA-Z_.]+}}, <3 x float> %{{[0-9a-zA-Z_.]+}})
+  x.a.x += 1;
+  return x;
+}
+
+// === TEST CASE: multi-level nested structs with single primitive type element at the bottom
+
+struct A1 { char x; };
+struct B1 { A1 a; };
+struct C1 {
+  B1 b;
+  C1 foo() { return *this; }
+};
+// CHECK-DAG: %struct.C1 = type { %struct.B1 }
+// CHECK-DAG: %struct.B1 = type { %struct.A1 }
+// CHECK-DAG: %struct.A1 = type { i8 }
+
+template SYCL_DEVICE C1 __regcall func<C1>(C1 x);
+// CHECK-DAG: define weak_odr x86_regcallcc i8 @_Z16__regcall3__funcI2C1ET_S1_(i8 %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: multi-level nested structs with multiple elements at all levels
+
+struct A2 { char x; };
+struct B2 { A2 a; int* ptr; };
+struct C2 {
+  B2 b;
+  double c;
+
+  C2 foo() { return *this; }
+};
+
+// CHECK-DAG: %struct.C2 = type { %struct.B2, double }
+// CHECK-DAG: %struct.B2 = type { %struct.A2, i32 addrspace(4)* }
+// CHECK-DAG: %struct.A2 = type { i8 }
+
+template SYCL_DEVICE C2 __regcall func<C2>(C2 x);
+// CHECK-DAG: define weak_odr x86_regcallcc %struct.C2 @_Z16__regcall3__funcI2C2ET_S1_(%struct.B2 %{{[0-9a-zA-Z_.]+}}, double %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: multi-level nested structs with one primitive type element at
+// the bottom, and one - at the top. The nested struct at the top is expected to
+// get "unwraped" by the compiler evaporating to the single element at the
+// bottom.
+
+struct A3 { char x; };
+struct B3 { A3 a; }; // unwrapped
+struct C3 { // unwrapped
+  B3 b;
+  char c;
+
+  C3 foo() { return *this; }
+};
+
+// CHECK-DAG: %struct.C3 = type { %struct.B3, i8 }
+// CHECK-DAG: %struct.B3 = type { %struct.A3 }
+// CHECK-DAG: %struct.A3 = type { i8 }
+
+template SYCL_DEVICE C3 __regcall func<C3>(C3 x);
+// CHECK-DAG: define weak_odr x86_regcallcc i16 @_Z16__regcall3__funcI2C3ET_S1_(i16 %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: multi-level nested structs with one primitive type element at
+// the bottom, and one - at the top. The nested struct at the top is expected to
+// get "unwraped" by the compiler evaporating to the single element at the
+// bottom.
```

Comment:
I think this comment is a copy-and-paste error describing the previous case.

Reply:
oh, right, will fix